components/suggestion/at_mention_provider/at_mention_provider.test.jsx

```diff
@@ -759,4 +760,126 @@ describe('components/suggestion/at_mention_provider/AtMentionProvider', () => {
             });
         });
     });
+
+    it('should prioritise username match over other matches for in channel users', async () => {
+        const pretext = '@x';
+        const matchedPretext = '@x';
+        const params = {
+            ...baseParams,
+            autocompleteUsersInChannel: jest.fn().mockImplementation(() => new Promise((resolve) => {
+                resolve({data: {
+                    users: [userid4, userid7],
+                    out_of_channel: [],
+                }});
+            })),
+        };
+
+        const provider = new AtMentionProvider(params);
+        const resultCallback = jest.fn();
+        expect(provider.handlePretextChanged(pretext, resultCallback)).toEqual(true);
+
+        expect(resultCallback).toHaveBeenNthCalledWith(1, {
+            matchedPretext,
+            terms: [
+                '@other',
+            ],
+            items: [
+                {type: Constants.MENTION_MEMBERS, ...userid3},
+            ],
+            component: AtMentionSuggestion,
+        });
+
+        jest.runAllTimers();
+
+        expect(resultCallback).toHaveBeenNthCalledWith(2, {
+            matchedPretext,
+            terms: [
+                '@other',
+                '',
+            ],
+            items: [
+                {type: Constants.MENTION_MEMBERS, ...userid3},
+                {type: Constants.MENTION_MORE_MEMBERS, loading: true},
+            ],
+            component: AtMentionSuggestion,
+        });
+
+        await Promise.resolve().then(() => {
```

Comment:
Instead of awaiting a new Promise here, couldn't we just call `jest.runAllTimers()` again between these tests?

Reply:
This doesn't work - `jest.runAllTimers()` doesn't flush the promise which triggers the 3rd call.

These tests copy the existing tests. I think the original tests may have used await Promise to around https://github.com/facebook/jest/issues/2157

Reply:
Ah, I see. Testing async code is kind of a mess then, but I guess this works